FirebaseMLModelDownloader/Sources/ModelDownloadTask.swift

```diff
@@ -43,33 +43,54 @@ class ModelDownloadTask: NSObject {
   private let appName: String
   /// Model info downloaded from server.
   private(set) var remoteModelInfo: RemoteModelInfo
+  /// Model conditions for download.
+  private let conditions: ModelDownloadConditions
   /// User defaults to which local model info should ultimately be written.
   private let defaults: UserDefaults
   /// Task to handle model file download.
   private var downloadTask: URLSessionDownloadTask?
   /// Progress and completion handlers associated with this model download task.
   private let downloadHandlers: DownloadHandlers
   /// Keeps track of download associated with this model download task.
-  private(set) var downloadStatus: ModelDownloadStatus = .notStarted
+  private(set) var downloadStatus: ModelDownloadStatus
+  /// Background session identifier.
+  private let backgroundSessionIdentifier = "ModelDownloadBackgroundSession"
+  /// Optional background completion handler.
+  private var backgroundCompletion: (() -> Void)?
   /// URLSession to handle model downloads.
-  private lazy var downloadSession = URLSession(configuration: .ephemeral,
-                                                delegate: self,
-                                                delegateQueue: nil)
+  private lazy var downloadSession: URLSession = {
+    var configuration = URLSessionConfiguration.ephemeral
+    if conditions.allowsBackgroundDownloading {
+      configuration = URLSessionConfiguration
+        .background(withIdentifier: backgroundSessionIdentifier)
```

Comment:
If I understand the [docs](https://developer.apple.com/documentation/foundation/url_loading_system/downloading_files_in_the_background) correctly, if the app is terminated while an ongoing download, then creating a url session with the same ID will associate the download tasks scheduled before termination with the session (see "Recreate the Session If the App Was Terminated" section). The code looks not ready to handle this case.